File: template_to_pdf/schema.py
def normalize(data):
    if not isinstance(data, dict):
        raise TypeError("expected object")
    title = data.get("title")
    if title is None:
        title = ""
    elif not isinstance(title, str):
        title = str(title)
    raw = data.get("sections", [])
    if not isinstance(raw, list):
        raise TypeError("sections must be a list")
    sections = []
    for item in raw:
        if not isinstance(item, dict):
            raise TypeError("section must be an object")
        heading = item.get("heading")
        body = item.get("body")
        sections.append({
            "heading": "" if heading is None else str(heading),
            "body": "" if body is None else str(body),
        })
    meta = data.get("meta") or {}
    if not isinstance(meta, dict):
        raise TypeError("meta must be an object")
    return {"title": title, "sections": sections, "meta": {str(k): "" if v is None else str(v) for k, v in meta.items()}}


def flatten(doc):
    doc = normalize(doc)
    vars_ = {"title": doc["title"]}
    chunks = [f"<h2>{s['heading']}</h2>\n<p>{s['body']}</p>" for s in doc["sections"]]
    vars_["body"] = "\n".join(chunks)
    for k, v in doc["meta"].items():
        vars_.setdefault(k, v)
    return vars_

File: js/schema.js
export function normalize(data) {
  if (data === null || typeof data !== "object" || Array.isArray(data)) {
    throw new TypeError("expected object")
  }
  let title = data.title
  if (title == null) title = ""
  else if (typeof title !== "string") title = String(title)
  let raw = data.sections
  if (raw == null) raw = []
  else if (!Array.isArray(raw)) throw new TypeError("sections must be a list")
  const sections = raw.map((item) => {
    if (item === null || typeof item !== "object" || Array.isArray(item)) {
      throw new TypeError("section must be an object")
    }
    return {
      heading: item.heading == null ? "" : String(item.heading),
      body: item.body == null ? "" : String(item.body),
    }
  })
  let meta = data.meta
  if (meta == null) meta = {}
  else if (typeof meta !== "object" || Array.isArray(meta)) {
    throw new TypeError("meta must be an object")
  }
  const outMeta = Object.fromEntries(
    Object.entries(meta).map(([k, v]) => [String(k), v == null ? "" : String(v)]),
  )
  return { title, sections, meta: outMeta }
}

export function flatten(doc) {
  doc = normalize(doc)
  const vars = { title: doc.title }
  vars.body = doc.sections.map((s) => `<h2>${s.heading}</h2>\n<p>${s.body}</p>`).join("\n")
  for (const [k, v] of Object.entries(doc.meta)) {
    if (!Object.prototype.hasOwnProperty.call(vars, k)) vars[k] = v
  }
  return vars
}
